Odata-docs/webapi-8/entityset-routing.md

```diff
@@ -0,0 +1,536 @@
+The partial structure of the controller for the OData service is as follows:
+
+```csharp
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.OData.Deltas;
+using Microsoft.AspNetCore.OData.Query;
+using Microsoft.AspNetCore.OData.Routing.Controllers;
+using EntitySetRouting.Models;
+
+public class ShapesController : ODataController
+{
+    private static List<Shape> shapes = new List<Shape>
+    {
+        new Rectangle { Id = 1, Length = 7, Width = 4, Area = 28 },
+        new Circle { Id = 2, Radius = 3.5, Area = 38.5 },
+        new Rectangle { Id = 3, Length = 8, Width = 5, Area = 40 }
+    };
+}
+```
+
+## Routing conventions for entity sets
+In this section we cover the conventions for entity set routing and the controller actions (endpoints) required for the requests to be routed successfully.
+
+### Retrieving an entity set
+The route template for this request is: `GET ~/{entityset}`.
+
+The following request returns the `Shapes` entity set - basically a collection of `Shape` entities:
+```http
+GET http://localhost:5000/odata/Shapes
+```
+
+For the above request to be conventionally-routed, a controller action named `Get` (or `GetShapes`) is expected:
+```csharp
+public ActionResult Get()
+{
+    return Ok(shapes);
+}
+```
+
+The following JSON payload shows the expected response:
+```json
+{
+    "@odata.context": "http://localhost:5000/odata/$metadata#Shapes",
+    "value": [
+        {
+            "@odata.type": "#EntitySetRouting.Models.Rectangle",
+            "Id": 1,
+            "Area": 28.0,
+            "Length": 7.0,
+            "Width": 4.0
+        },
+        {
+            "@odata.type": "#EntitySetRouting.Models.Circle",
+            "Id": 2,
+            "Area": 38.5,
+            "Radius": 3.5
+        },
+        {
+            "@odata.type": "#EntitySetRouting.Models.Rectangle",
+            "Id": 3,
+            "Area": 40.0,
+            "Length": 8.0,
+            "Width": 5.0
+        }
+    ]
+}
+```
+
+The response contains 3 shape objects - 2 rectangles and 1 circle. Since `Rectangle` and `Circle` are derived types, each of the shape objects contain an `@odata.type` property specifying the type of the entity.
+
+### Retrieving the count of an entity set
+The route template for this request is: `GET ~/{entityset}/$count`.
+
+To address the raw value of the number of items in an entity set, append `/$count` to that entity set's URL.
+
+The following request returns a count of items in the `Shapes` entity set:
+```http
+GET http://localhost:5000/odata/Shapes/$count
+```
+
+For the above request to be conventionally-routed, a controller action named `Get` (or `GetShapes`) is expected, same as is expected when addressing an entity set. However, the controller action needs to be decorated with `EnableQuery` attribute. The `EnableQuery` attribute is responsible of generating the relevant query for determining the count of items:
```

Comment:
If you had enabled the query options when setting up the model do you still have to decorate the controller action methods with `EnableQuery`

Reply:
@ElizabethOkerio In _Startup.cs_ we only enable the features. It's the `EnableQuery` attribute that does the magic. It is required. Alternatively, someone can add the `ODataQueryOptions<T>` parameter to the method and then do `options.ApplyTo(IQueryable)` in the method body. This alternative way should be covered in the OData Query Options doc. I didn't want to introduce it in the context of routing to avoid confusion

Reply:
What are the features? Also you can do something like .Count().Expand.Select(). This is just for my own knowledge.

Reply:
@ElizabethOkerio Yeah... `.Count().Expand.Select()` are the OData query features I'm talking about here. We actually support a method `EnableQueryFeatures(int?)` that enables them all instead of doing `.Select().Filter().OrderBy().Expand().Count().SkipToken().SetMaxTop(null)`. I have used that method in all the doc pages in this PR to keep the statement short